server/src/main/java/org/elasticsearch/action/search/QueryPhaseResultConsumer.java

```diff
@@ -277,10 +347,20 @@ private MergeResult partialReduce(
                 }
             }
             // we have to merge here in the same way we collect on a shard
-            newTopDocs = topDocsList == null ? null : mergeTopDocs(topDocsList, topNSize, 0);
-            newAggs = hasAggs
-                ? aggregate(toConsume.iterator(), lastMerge, resultSetSize, aggReduceContextBuilder.forPartialReduction())
-                : null;
+            newTopDocs = topDocsList == null ? Lucene.EMPTY_TOP_DOCS : mergeTopDocs(topDocsList, topNSize, 0);
+            if (hasAggs) {
+                newAggs = aggregate(
+                    toConsume.iterator(),
+                    lastMerge == null ? Collections.emptyIterator() : Iterators.single(lastMerge.reducedAggs),
+                    resultSetSize,
+                    aggReduceContextBuilder.forPartialReduction()
+                );
+            } else {
+                newAggs = null;
+            }
+            for (QuerySearchResult querySearchResult : toConsume) {
+                querySearchResult.markAsPartiallyReduced();
+            }
```

Comment:
here too, I am a bit confused about the changes made. Could you keep down to the minimum required?

Reply:
This is the shortest version I could find sorry, again, this became a little trickier with https://github.com/elastic/elasticsearch/pull/124520 but there's just no completely trivial ways to keep heap usage flat I think, you need to intermix working and releasing pieces.

Reply:
I revered that one ternary I adjusted, the other changes are either needed to not have nulls for serialization (the EMPTY_TOP_DOCS) or come out of the O(1) memory changes bringing back the aggslist as a short term trivial solution to incremental dereferencing.